### PR TITLE
handles AwaitAsync finishing synchronously

### PR DESCRIPTION
**Pull request by gregoryyoung**
Affects all HttpListener usages.

A connect flood will break the async loop and take down the entire HttpListener instance (will stop accepting connections).

To reproduce. Take a HttpListener run say 2 wrk test instances against it (1000 connections each works here). After the run the HttpListener has broken its async loop and will no longer accept any connections after. The reason for this is that it was not previously handling the case that AcceptAsync can return synchronously. In this case event will not be called with args.

Patch handles this case and the above issue is not reproducible with.

### CODE REVIEW

**Comment by akoeplinger** [comment]
whitelist

**Comment by akoeplinger** [comment]
The code changes look good to me. Could you please make the following changes:
1. Use tabs, not spaces
2. space between method name and opening parenthesis (Mono coding convention)

**Comment by kumpera** [comment]
@ludovic-henry can you review this one?

**Comment by knocte** [comment]
Also please paste the description of the pull request in your commit message. Your current commit message is too short.

File: mcs/class/System/System.Net/EndPointListener.cs [hunk 1]
```diff
@@ -73,7 +73,7 @@ public EndPointListener (HttpListener listener, IPAddress addr, int port, bool s
 			SocketAsyncEventArgs args = new SocketAsyncEventArgs ();
 			args.UserToken = this;
 			args.Completed += OnAccept;
-			sock.AcceptAsync (args);
+                        Accept (sock, args);
```
Comment (ludovic-henry) [inline mcs/class/System/System.Net/EndPointListener.cs]: All your added code is indented one space too much.

File: mcs/class/System/System.Net/EndPointListener.cs [hunk 2]
```diff
@@ -82,28 +82,26 @@ public EndPointListener (HttpListener listener, IPAddress addr, int port, bool s
...
+                static void ProcessAccept (SocketAsyncEventArgs args)
+                {
+                        Socket accepted = null;
+                        if (args.SocketError == SocketError.Success)
+                                accepted = args.AcceptSocket;
+
+                        EndPointListener epl = (EndPointListener) args.UserToken;
+
+
+                        Accept (epl.sock, args);
```
Comment (ludovic-henry) [inline mcs/class/System/System.Net/EndPointListener.cs]: We are missing the `try { ... } catch { ...; accepted = null; }` from before.

**Comment by ludovic-henry** [comment]
LGTM

File: mcs/class/System/System.Net/EndPointListener.cs (hunk 2 above)
Comment (gregoryyoung) [inline mcs/class/System/System.Net/EndPointListener.cs]: What are you hoping that code to do? Your entire listener is hosed at that
point (similar to the issue this patch resolves) as you have broken the
async loop, no future accept will happen. Accepted may also still be a
valid socket. I am not clear what the benefit of setting the accepted
socket to null would be.

On Mon, Mar 7, 2016 at 2:11 PM, Ludovic Henry <notifications@github.com>
wrote:

> In mcs/class/System/System.Net/EndPointListener.cs
> <https://github.com/mono/mono/pull/2713#discussion_r55194917>:
>
> >
> > +                static void ProcessAccept (SocketAsyncEventArgs args)
> > +                {
> > +                        Socket accepted = null;
> > +                        if (args.SocketError == SocketError.Success)
> > +                                accepted = args.AcceptSocket;
> > +
> > +                        EndPointListener epl = (EndPointListener) args.UserToken;
> > +
> > +
> > +                        Accept (epl.sock, args);
>
> We are missing the try { ... } catch { ...; accepted = null; } from
> before.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mono/mono/pull/2713/files#r55194917>.
>



-- 
Studying for the Turing test

**Comment by gregoryyoung** [comment]
Changed formatting. Squashed. Put more details in merged commit.

**Comment by gregoryyoung** [comment]
@ludovic-henry See comment above. What is the purpose of the code I didn't bring over? I don't mind bringing it over but don't really understand why its there.

Overall we just had someone else just hit this same issue this week. They were working from go and didn't use keep alives. 10k requests took down http listener.

**Comment by ludovic-henry** [comment]
@gregoryyoung fair enough. As soon as the testing is green we can merge this PR. Also, could you please squash the commits together, as @akoeplinger noted, they are not. Thank you!

**Comment by ludovic-henry** [comment]
build

**Comment by lewurm** [comment]
@gregoryyoung can you squash your changes as suggested please? then it should be ready to merge.

**Comment by gregoryyoung** [comment]
done.

**Comment by esdrubal** [comment]
@alexrp pointed out on exception that shouldn't happen before this pull request:
```
E/mono    ( 2367): System.ObjectDisposedException: Cannot access a disposed object.
E/mono    ( 2367): Object name: 'System.Net.Sockets.Socket'.
E/mono    ( 2367):   at System.Net.Sockets.Socket.ThrowIfDisposedAndClosed () <0xa0e7fa08 + 0x0004b> in <filename unknown>:0 
E/mono    ( 2367):   at System.Net.Sockets.Socket.AcceptAsync (System.Net.Sockets.SocketAsyncEventArgs e) <0xa0e81728 + 0x00017> in <filename unknown>:0 
E/mono    ( 2367):   at System.Net.EndPointListener.Accept (System.Net.Sockets.Socket socket, System.Net.Sockets.SocketAsyncEventArgs e) <0xa0e816b0 + 0x00043> in <filename unknown>:0 
E/mono    ( 2367):   at System.Net.EndPointListener.ProcessAccept (System.Net.Sockets.SocketAsyncEventArgs args) <0xa0e8a168 + 0x00063> in <filename unknown>:0 
E/mono    ( 2367):   at System.Net.EndPointListener.OnAccept (System.Object sender, System.Net.Sockets.SocketAsyncEventArgs e) <0xa0e8a150 + 0x00013> in <filename unknown>:0 
E/mono    ( 2367):   at System.Net.Sockets.SocketAsyncEventArgs.OnCompleted (System.Net.Sockets.SocketAsyncEventArgs e) <0xa0e8a118 + 0x00028> in <filename unknown>:0 
E/mono    ( 2367):   at System.Net.Sockets.SocketAsyncEventArgs.Complete () <0xa0e8a100 + 0x00014> in <filename unknown>:0 
E/mono    ( 2367):   at System.Net.Sockets.Socket.<AcceptAsyncCallback>m__0 (System.IAsyncResult ares) <0xa0e89b40 + 0x002b3> in <filename unknown>:0 
E/mono    ( 2367):   at System.Net.Sockets.SocketAsyncResult+<Complete>c__AnonStorey0.<>m__0 (System.Object _) <0xa0e89b18 + 0x0001b> in <filename unknown>:0 
E/mono    ( 2367):   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () <0xa1c2db70 + 0x00028> in <filename unknown>:0 
E/mono    ( 2367):   at System.Threading.ThreadPoolWorkQueue.Dispatch () <0xa0ecd9d0 + 0x00185> in <filename unknown>:0 
E/mono    ( 2367):   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () <0xa0ecd8c8 + 0x0000b> in <filename unknown>:0 
```
We probably need to check that the socket was not disposed before calling Accept in ProcessAccept.

The exception occurred on wrench monodroid test suite.

**Comment by akoeplinger** [comment]
@esdrubal you're only seeing this in master, not in C7 right?

**Comment by esdrubal** [comment]
@alexrp pointed out a crash on monodroid master. I don't know of other instances of it.

**Pull request merged**